google/resource_kms_crypto_key.go

```diff
@@ -0,0 +1,265 @@
+		return err
+	}
+
+	log.Printf("[WARNING] KMS CryptoKey resources cannot be deleted from GCP. This CryptoKey %s will be removed from Terraform state, but will still be present on the server.", cryptoKeyId.cryptoKeyId())
+
+	d.SetId("")
+
+	err = clearCryptoKeyVersions(cryptoKeyId, config)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateKmsCryptoKeyRotationPeriod(period string) error {
+	pattern := regexp.MustCompile("^([0-9.]*\\d)s$")
+	match := pattern.FindStringSubmatch(period)
+
+	if len(match) == 0 {
+		return fmt.Errorf("Invalid period format: %s", period)
+	}
+
+	number := match[1]
+	seconds, err := strconv.ParseFloat(number, 64)
+
+	if err == nil && seconds < 86400.0 {
+		return fmt.Errorf("Rotation period must be greater than one day")
+	}
+
+	parts := strings.Split(number, ".")
+
+	if err == nil && len(parts) > 1 && len(parts[1]) > 9 {
+		return fmt.Errorf("Rotation period cannot have more than 9 fractional digits")
+	}
+
+	return nil
+}
+
+func kmsCryptoKeyNextRotation(now time.Time, period string) (string, error) {
+	var result string
+	var duration time.Duration
+
+	err := validateKmsCryptoKeyRotationPeriod(period)
+
+	if err == nil {
+		duration, err = time.ParseDuration(period)
+	}
+
+	if err == nil {
+		result = now.UTC().Add(duration).Format(time.RFC3339Nano)
+	}
+
+	return result, err
+}
+
+func parseKmsCryptoKeyId(id string, config *Config) (*kmsCryptoKeyId, error) {
+	parts := strings.Split(id, "/")
+
+	cryptoKeyIdRegex := regexp.MustCompile("^([a-z0-9-]+)/([a-z0-9-])+/([a-zA-Z0-9_-]{1,63})+/([a-zA-Z0-9_-]{1,63})$")
+	cryptoKeyIdWithoutProjectRegex := regexp.MustCompile("^([a-z0-9-])+/([a-zA-Z0-9_-]{1,63})+/([a-zA-Z0-9_-]{1,63})$")
+
+	if cryptoKeyIdRegex.MatchString(id) {
+		return &kmsCryptoKeyId{
+			KeyRingId: kmsKeyRingId{
+				Project:  parts[0],
+				Location: parts[1],
+				Name:     parts[2],
+			},
+			Name: parts[3],
+		}, nil
+	}
+
+	if cryptoKeyIdWithoutProjectRegex.MatchString(id) {
+		if config.Project == "" {
+			return nil, fmt.Errorf("The default project for the provider must be set when using the `{location}/{cryptoKeyName}` id format.")
+		}
+
+		return &kmsCryptoKeyId{
+			KeyRingId: kmsKeyRingId{
+				Project:  config.Project,
+				Location: parts[0],
+				Name:     parts[1],
+			},
+			Name: parts[2],
+		}, nil
+	}
+
+	return nil, fmt.Errorf("Invalid CryptoKey id format, expecting `{projectId}/{locationId}/{KeyringName}/{cryptoKeyName}` or `{locationId}/{KeyringName}/{cryptoKeyName}.`")
+}
+
+func resourceKmsCryptoKeyImportState(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	config := meta.(*Config)
+
+	cryptoKeyId, err := parseKmsCryptoKeyId(d.Id(), config)
+	if err != nil {
+		return nil, err
+	}
+
+	d.Set("key_ring", cryptoKeyId.KeyRingId.terraformId())
+	d.Set("name", cryptoKeyId.Name)
+
+	if d.Get("purpose") == "" {
+		d.Set("purpose", "ENCRYPT_DECRYPT")
+	}
+
+	d.SetId(cryptoKeyId.terraformId())
+
+	return []*schema.ResourceData{d}, nil
+}
```

Comment:
It might be a good idea to set `rotation_period` if it exists.

Reply:
Done - further up so comment isn't hidden ...